graph/src/util/lfu_cache.rs

```diff
@@ -0,0 +1,192 @@
+                        will_stale: false,
+                    },
+                    (false, Reverse(1)),
+                );
+            }
+            Some(entry) => {
+                entry.weight = weight;
+                entry.value = value;
+                entry.will_stale = false;
+                self.total_weight += weight - entry.weight;
+            }
+        }
+    }
+
+    fn get_mut(&mut self, key: K) -> Option<&mut CacheEntry<K, V>> {
+        // Increment the frequency by 1
+        let key_entry = CacheEntry::cache_key(key);
+        self.queue
+            .change_priority_by(&key_entry, |(s, Reverse(f))| (s, Reverse(f + 1)));
+        self.queue.get_mut(&key_entry).map(|x| x.0)
+    }
+
+    pub fn get(&mut self, key: &K) -> Option<&V> {
+        self.get_mut(key.clone()).map(|x| &x.value)
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        // `PriorityQueue` doesn't have a remove method, so emulate that by setting the priority to
+        // the absolute minimum and popping.
+        let key_entry = CacheEntry::cache_key(key.clone());
+        self.queue
+            .change_priority(&key_entry, (true, Reverse(u64::min_value())))
+            .and_then(|_| {
+                self.queue.pop().map(|(e, _)| {
+                    assert_eq!(e.key, key_entry.key);
+                    self.total_weight -= e.weight;
+                    e.value
+                })
+            })
+    }
+
+    pub fn contains_key(&self, key: &K) -> bool {
+        self.queue
+            .get(&CacheEntry::cache_key(key.clone()))
+            .is_some()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.queue.is_empty()
+    }
+
+    pub fn evict(&mut self, max_weight: u64) {
+        if self.total_weight <= max_weight {
+            return;
+        }
+
+        self.stale_counter += 1;
+        if self.stale_counter == STALE_PERIOD {
+            self.stale_counter = 0;
+
+            // Entries marked `will_stale` are effectively set as stale.
+            // All entries are then marked `will_stale`.
+            for (e, p) in self.queue.iter_mut() {
+                p.0 = e.will_stale;
+                e.will_stale = true;
+            }
+        }
+
+        while self.total_weight > max_weight {
+            // Unwrap: If there were no entries left, `total_weight` would be 0.
+            let entry = self.queue.pop().unwrap().0;
+            self.total_weight -= entry.weight;
+        }
+    }
+}
+
+impl<K: Ord + Eq + Hash + 'static, V: 'static> IntoIterator for LfuCache<K, V> {
+    type Item = (CacheEntry<K, V>, Priority);
+    type IntoIter = Box<dyn Iterator<Item = Self::Item>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Box::new(self.queue.into_iter())
+    }
+}
+
+impl<K: Ord + Eq + Hash, V> Extend<(CacheEntry<K, V>, Priority)> for LfuCache<K, V> {
+    fn extend<T: IntoIterator<Item = (CacheEntry<K, V>, Priority)>>(&mut self, iter: T) {
+        self.queue.extend(iter);
+    }
+}
```

Comment:
The `LfuCache` deserves unit tests, wouldn't you agree? 😉

Reply:
Yes! I've added one.